### PR TITLE
[java] Fix #4435 - FP with UnusedAssignment

### LINKED ISSUE #3694: [java] SingularField ignores static variables
**Affects PMD Version:**
6.41.0+
7.0.0 

**Rule:SingularField**

Please provide the rule name and a link to the rule documentation:
<https://pmd.github.io/latest/pmd_rules_java_design.html#singularfield>

**Description:**
Hi, I found two false negatives about the rule SingularField. Please refer to the following case, PMD should have reported a false negative at line 2 and 3. I think these cases are caused by modifiers of `var1` and `var2`.

**Code Sample demonstrating the issue:**
```java
class Impl {
  final private int var1 = 10;  // should report a warning here
  static private int var2 = 10; // also should report a warning here
  public int foo() {
    return var1;
  }
  public int foo(int y) {
    var2 = y + 5;
    return var2;
  }
}
```

**Expected outcome:**

PMD should report violations at line 2 and 3, but doesn't. This is a false-negative.

**Running PMD through:** *[Maven]*

### LINKED ISSUE #4435: [java] [7.0-rc1] UnusedAssignment for used field
**Affects PMD Version:** 
7.0-rc1 when upgrading from 6.55

**Rule:**
[UnusedAssignment](https://pmd.sourceforge.io/pmd-7.0.0-rc1/pmd_rules_java_bestpractices.html#unusedassignment)

**Description:**
A class field is not the same as a method variable as it is in scope for other method calls. This also only restores the value when an unhandled error occurs to rollback the state, so the detection is misunderstanding the behavior.

**Code Sample demonstrating the issue:**
```java
public void advance(BoundedLocalCache<K, V> cache, long currentTimeNanos) {
  long previousTimeNanos = nanos;
  nanos = currentTimeNanos;

  try {
    // ...
  } catch (Throwable t) {
    nanos = previousTimeNanos;
    throw t;
  }
}
```

**Expected outcome:**
PMD reports a violation but that should not apply to a class field.

\# | File | Line | Problem
-- | -- | -- | --
1 | TimerWheel.java | 90 | The value assigned to field 'nanos' is never used (overwritten on line 111)

**Running PMD through:** *Gradle*
This requires workaround in https://github.com/gradle/gradle/issues/24502

### LINKED ISSUE #4618: [java] UnusedAssignment false positive with conditional assignments of fields
**Affects PMD Version:** 
7.0.0-SNAPSHOT

**Rule:**

[https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#unusedassignment](UnusedAssignment)

**Description:**

**Code Sample demonstrating the issue:**

```java
Selftest.java
```

**Expected outcome:**

PMD reports violations, but that's wrong. These are false positives:
```
Selftest.java:162:	UnusedAssignment:	The value assigned to field 'withIssues' is never used (overwritten on line 168)
Selftest.java:177:	UnusedAssignment:	The value assigned to field 'withIssues' is never used (overwritten on line 184)
Selftest.java:666:	UnusedAssignment:	The value assigned to field 'modified' is never used (overwritten on lines 682 and 703)
Selftest.java:703:	UnusedAssignment:	The value assigned to field 'modified' is never used (reassigned every iteration)
Selftest.java:704:	UnusedAssignment:	The value assigned to field 'withIssues' is never used (overwritten on lines 704 and 762)
```

The values are assigned conditionally, within an if-clause.

**Running PMD through:** CLI/Maven

Files attached.
[file-and-ruleset3.zip](https://github.com/pmd/pmd/files/11909818/file-and-ruleset3.zip)

### PR DESCRIPTION
**Pull request by oowekyala**
## Related issues

- Fix #4435 
- Fix #3694
- Fix #4618 

## Ready?

- [x] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 6 violations,<br />
introduces 757 new violations, 0 new errors and 0 new configuration errors,<br />
removes 247 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/3b33d95eb3214d2c97f1c5277727a456/diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by oowekyala** [comment]
The regression tester comment above seems broken:
```html
<a href="Chunk.io now requires authentication. Send an email to hello@chunk.io to get an invite./diff1/index.html">Full report</a></td>
```

**Comment by adangel** [comment]
> The regression tester comment above seems broken:
> 
> ```
> <a href="Chunk.io now requires authentication. Send an email to hello@chunk.io to get an invite./diff1/index.html">Full report</a></td>
> ```

Yes, it seems we lost the account there (probably due to inactivity). I've changed the github actions script (https://github.com/pmd/pmd/commit/d238de53071a28ff74609dad2c3779f1bd2e8387 and https://github.com/pmd/pmd/commit/f5829ac5b15fb410ef65040e3334339319f6e1bc), so that the regression tester report is uploaded as an artifact. It worked for the other pull request already: #4605 -> go to checks -> left to the button "re-run all jobs" there is "artifacts" where you can download the report. You can view then the report locally.

You'll need to merge/rebase against master to get the new github actions script. With the next build, you should also see this artifact then.

**Comment by oowekyala** [comment]
Thanks for explaining!

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/internal/DataflowPass.java [hunk 1]
```diff
@@ -894,15 +892,14 @@ private SpanInfo processAssignment(ASTExpression lhs, // LHS or unary operand
         }
 
         private boolean isRelevantField(ASTExpression lhs) {
-            return lhs instanceof ASTNamedReferenceExpr
-                && (trackThisInstance() && JavaAstUtils.isThisFieldAccess(lhs)
-                || trackStaticFields() && isStaticFieldOfThisClass(((ASTNamedReferenceExpr) lhs).getReferencedSym()));
+            return lhs instanceof ASTNamedReferenceExpr && (trackThisInstance() && JavaAstUtils.isThisFieldAccess(lhs)
+                || true && isStaticFieldOfThisClass(((ASTNamedReferenceExpr) lhs).getReferencedSym()));
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/internal/DataflowPass.java:896]: this `|| true` seems to have slipped past some local experiments…

**Comment by oowekyala** [comment]
After reviewing the diff report, this seems alright to me. It fixes some unreported FNs/FPs on ImmutableField too.

**Review by adangel — APPROVED** [review]
Thanks!

**Pull request merged**